File: c#/kmurrayDay1.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace AdventOfCode2016
{
   internal static class Day1
   {
      private enum Direction
      {
         North,
         East,
         South,
         West
      };

      public static void Process()
      {
         var xposition = 0;
         var yposition = 0;
         var vectors = new HashSet<string>();

         FindPosition(ref xposition, ref yposition, vectors);

         Console.WriteLine("Ending position x = " + xposition + " y = " + yposition);
         Console.WriteLine("Resulting distance from the Easter Bunny = " + (Math.Abs(xposition) + Math.Abs(yposition)));
      }

      private static void AddPoint(int xpos, int ypos, int move, int xmove, HashSet<String> vectors) {

         var direction = (move < 0) ? -1 : 1;

         for (var count = 0; count < Math.Abs(move) ; ++count) {
            switch (xmove)
            {
                case 0 :
                    if (vectors.Contains(xpos + "," + (ypos + (count * direction)))) {
                        Console.WriteLine("Intersection at " + xpos + "," + (ypos + (count * direction)) + " Distance from home = " + (Math.Abs(xpos) + Math.Abs(ypos + count)));
                    }
                    vectors.Add(xpos + "," + (ypos + count));
                    break;
                case 1 :
                    if (vectors.Contains((xpos + (count * direction)) + "," + ypos))
                    {
                        Console.WriteLine("Intersection at " + (xpos + (count * direction)) + "," + ypos + " Distance from home = " + (Math.Abs(xpos + count) + Math.Abs(ypos)));
                    }
                    vectors.Add((xpos + (count * direction)) + "," + ypos);
                    break;
            }
         }
      }

      private static void FindPosition(ref int xpos, ref int ypos, HashSet<String> vectors) {
         var inputFile = File.ReadAllText(Path.Combine("Inputs", "Day1a.txt")).Replace("\r", "").Replace("\n", "").Replace(" ", "").Split(',');
         var position = Direction.North;
         
         foreach (var move in inputFile) {

            if (move.Substring(0, 1).Equals("R", StringComparison.InvariantCultureIgnoreCase)) {

               switch (position) {
                  case Direction.North:
                     AddPoint(xpos, ypos, Convert.ToInt32(move.Substring(1, move.Length - 1)), 1, vectors);
                     xpos += Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.East;
                     break;
                  case Direction.East:
                     AddPoint(xpos, ypos, -(Convert.ToInt32(move.Substring(1, move.Length - 1))), 0, vectors);
                     ypos -= Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.South;
                     break;
                  case Direction.South:
                     AddPoint(xpos, ypos, -(Convert.ToInt32(move.Substring(1, move.Length - 1))), 1, vectors);
                     xpos -= Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.West;
                     break;
                  case Direction.West:
                     AddPoint(xpos, ypos, Convert.ToInt32(move.Substring(1, move.Length - 1)), 0, vectors);
                     ypos += Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.North;
                     break;
               }
            }
            else if (move.Substring(0, 1).Equals("L", StringComparison.CurrentCultureIgnoreCase)) {
               switch (position) {
                  case Direction.North:
                     AddPoint(xpos, ypos, -(Convert.ToInt32(move.Substring(1, move.Length - 1))), 1, vectors);
                     xpos -= Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.West;
                     break;
                  case Direction.East:
                     AddPoint(xpos, ypos, Convert.ToInt32(move.Substring(1, move.Length - 1)), 0, vectors);
                     ypos += Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.North;
                     break;
                  case Direction.South:
                     AddPoint(xpos, ypos, Convert.ToInt32(move.Substring(1, move.Length - 1)), 1, vectors);
                     xpos += Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.East;
                     break;
                  case Direction.West:
                     AddPoint(xpos, ypos, -(Convert.ToInt32(move.Substring(1, move.Length - 1))), 0, vectors);
                     ypos -= Convert.ToInt32(move.Substring(1, move.Length - 1));
                     position = Direction.South;
                     break;
               }
            }
         }
      }
   }
}
/*
--- Day 1: No Time for a Taxicab ---

Santa's sleigh uses a very high-precision clock to guide its movements, and the clock's oscillator is regulated by stars.Unfortunately, the stars have been stolen...by the Easter Bunny.To save Christmas, Santa needs you to retrieve all fifty stars by December 25th.

Collect stars by solving puzzles.Two puzzles will be made available on each day in the advent calendar; the second puzzle is unlocked when you complete the first.Each puzzle grants one star. Good luck!


You're airdropped near Easter Bunny Headquarters in a city somewhere. "Near", unfortunately, is as close as you can get - the instructions on the Easter Bunny Recruiting Document the Elves intercepted start here, and nobody had time to work them out further.


The Document indicates that you should start at the given coordinates (where you just landed) and face North.Then, follow the provided sequence: either turn left(L) or right(R) 90 degrees, then walk forward the given number of blocks, ending at a new intersection.

There's no time to follow such ridiculous instructions on foot, though, so you take a moment and work out the destination. Given that you can only walk on the street grid of the city, how far is the shortest path to the destination?

For example:

Following R2, L3 leaves you 2 blocks East and 3 blocks North, or 5 blocks away.
R2, R2, R2 leaves you 2 blocks due South of your starting position, which is 2 blocks away.
R5, L5, R5, R3 leaves you 12 blocks away.
How many blocks away is Easter Bunny HQ?

Your puzzle answer was 301.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---

Then, you notice the instructions continue on the back of the Recruiting Document.Easter Bunny HQ is actually at the first location you visit twice.

For example, if your instructions are R8, R4, R4, R8, the first location you visit twice is 4 blocks away, due East.

How many blocks away is the first location you visit twice? 

Puzzle Answer is 130.
*/

File: swift/kmurrayDay1.swift
//
// Created by Kevin Murray on 12/6/16.
// Copyright (c) 2016 Kevin Murray. All rights reserved.
//

import Foundation

class Day1 {
   enum Direction: Int {
      case North = 0
      case East = 1
      case South = 2
      case West = 3
   }
   static var firstIntersection = 0;
   static var intersection = "";

   public static func process() -> String {
      var xpos = 0;
      var ypos = 0;
      let vectors = NSMutableSet();

      findPosition(xpos: &xpos, ypos: &ypos, vectors: vectors);

      print("Ending position x = \(xpos) y = \(ypos)");
      let distance = Swift.abs(xpos) + Swift.abs(ypos);
      print("Resulting distance from Easter Bunny = \(distance)");

      var result = "Ending position x = \(xpos) y = \(ypos) distance = \(distance)\n";
      result += "\(intersection)\n";
      return result;
   }

   static func findPosition(xpos: inout Int, ypos: inout Int, vectors: NSMutableSet) {
      var position = Direction.North;

      do {
         var data = try String(contentsOfFile: "/Users/kmurray/src/aoc/aoc/Day1.txt", encoding: String.Encoding.utf8).replacingOccurrences(of: " ", with: "");

         let moveArr = data.characters.split{$0 == ","}.map(String.init);


         for move in moveArr {

            let dIndex = move.index(move.startIndex, offsetBy: 1);
            let dMove = Int(move.substring(from: dIndex));

            if (move.substring(to: dIndex) == "R") {

               switch position {
               case Direction.North:
                  AddPoint(xpos: xpos, ypos: ypos, move: dMove!, xMove: 1, vectors: vectors);
                  xpos += dMove!;
                  position = Direction.East;
               case Direction.East:
                  AddPoint(xpos: xpos, ypos: ypos, move: -dMove!, xMove: 0, vectors: vectors);
                  ypos -= dMove!;
                  position = Direction.South;
               case Direction.South:
                  AddPoint(xpos: xpos, ypos: ypos, move: -dMove!, xMove: 1, vectors: vectors);
                  xpos -= dMove!;
                  position = Direction.West;
               case Direction.West:
                  AddPoint(xpos: xpos, ypos: ypos, move: dMove!, xMove: 0, vectors: vectors);
                  ypos += dMove!;
                  position = Direction.North;
               }
            }

            else if (move.substring(to: dIndex) == "L") {
               switch position {
               case Direction.North:
                  AddPoint(xpos: xpos, ypos: ypos, move: -dMove!, xMove: 1, vectors: vectors);
                  xpos -= dMove!;
                  position = Direction.West;
               case Direction.East:
                  AddPoint(xpos: xpos, ypos: ypos, move: dMove!, xMove: 0, vectors: vectors);
                  ypos += dMove!;
                  position = Direction.North;
               case Direction.South:
                  AddPoint(xpos: xpos, ypos: ypos, move: dMove!, xMove: 1, vectors: vectors);
                  xpos += dMove!;
                  position = Direction.East;
               case Direction.West:
                  AddPoint(xpos: xpos, ypos: ypos, move: -dMove!, xMove: 0, vectors: vectors);
                  ypos -= dMove!;
                  position = Direction.South;
               }
            }
         }
      }
      catch {}

   }

   static func AddPoint(xpos: Int, ypos: Int, move: Int, xMove: Int, vectors: NSMutableSet) {
      var count = 0;
      let direction = (move < 0) ? -1 : 1;

      if (firstIntersection == 0) {
         while (count < Swift.abs(move)) {
            if (xMove == 0) {
               let index = "\(xpos),\(ypos + (count * direction))";

               if (vectors.contains(index)) {
                  let distance = Swift.abs(xpos) + Swift.abs(ypos + count);
                  intersection = "Intersection at \(index) distace from home = \(distance)";
                  firstIntersection = 1;
               }
               vectors.add(index);
            } else if (xMove == 1) {
               let index = "\(xpos + (count * direction)),\(ypos)";
               if (vectors.contains(index)) {
                  let distance = Swift.abs(xpos + count) + Swift.abs(ypos);
                  intersection = "Intersection2 at \(index) distance from home = \(distance)";
                  firstIntersection = 1;
               }
               vectors.add(index);
            }
            count += 1;
         }
      }
   }
}
